open-apparel-registry/open-apparel-registry
# Add Terraform configuration for staging environment

## Overview

* Bootstrap `deployment/terraform` directory structure and initialize Terraform S3 backend
* Add `infra` script
* Using https://github.com/azavea/terraform-aws-ecr-repository, instantiate an ECR repository
* Add `cipublish` script that publishes container images to ECR
* Add support for `cipublish` and `infra` stages in `Jenkinsfile` (gated to develop and release branches)
* ECS with Fargate
* Include VPC with [terraform-aws-vpc](https://github.com/azavea/terraform-aws-vpc)
* Include PostgreSQL RDS with [terraform-aws-postgresql-rds](https://github.com/azavea/terraform-aws-postgresql-rds)
* Include log aggregation with CloudWatch Logs
* Provision ACM certificates with [terraform-aws-acm-certificate](https://github.com/azavea/terraform-aws-acm-certificate)
* Layer in CloudFront distribution to front ALB
* Install `requests` so we can query the ECS Task Metadata endpoint to allow health checks to pass
* Thread `GOOGLE_GEOCODING_API_KEY` through to ECS task definition

Resolves #58 
Resolves #59 

## Notes

For accessing the Bastion, I generated a set of  `oar-stg` and `oar-prd` key pairs. They live in marriage rings.

I also had to copy the `NS` records from the `staging.openapparel.org` hosted zone into the `openapparel.org` hosted zone, during the Terraform cycle, in order for the ACM validation to go through.

I created an `OAR_AWS_ECR_ENDPOINT` credential on `civicci01`.
 
I think we should be OK with the default resource allocation for now (0.25 vCPU and 512MB of memory).

You'll probably want to take a look at the values set in the `terraform.tfvars` file on S3 as part of reviewing this PR.

## Testing Instructions

See that Terraform plan is clean:

```bash
vagrant@vagrant:/vagrant$ docker-compose -f docker-compose.ci.yml run --rm terraform
bash-4.4# export GIT_COMMIT=4d3e21f
bash-4.4# ./scripts/infra plan
...
------------------------------------------------------------------------

No changes. Infrastructure is up-to-date.

This means that Terraform did not detect any differences between your
configuration and real physical resources that exist. As a result, no
actions need to be performed.
+ popd
/usr/local/src
bash-4.4#
```

Visit https://staging.openapparel.org/
